modules/ocf_mirrors/files/collect-mirrors-stats

```diff
@@ -0,0 +1,168 @@
+import os
+import traceback
+from datetime import date
+from datetime import datetime
+from datetime import timedelta
+from pathlib import Path
+
+from ocflib.lab.stats import get_connection
+from ocflib.lab.stats import humanize_bytes
+
+OCFSTATS_PWD = os.environ.get('OCFSTATS_PWD', None)
+MIRRORS_DATA_PATH = Path('/opt/mirrors/ftp')
+APACHE_LOG_PATH = Path('/var/log/apache2')
+APACHE_LOG_FILES = [
+    Path('mirrors.ocf.berkeley.edu_access.log'),
+    Path('mirrors.ocf.berkeley.edu-ssl_access_ssl.log')
+]
+RSYNC_LOG_PATH = Path('/var/log/rsync')
+
+
+def log_to_mysql(projects, dt=None, quiet=False):
+    log_date = dt or date.today()
+
+    with get_connection(user='ocfstats', password=OCFSTATS_PWD) as cursor:
+        for name, data in projects.items():
+            cursor.execute(
+                'INSERT INTO `mirrors` (`date`, `dist`, `up`, `down`) '
+                'VALUES (%s, %s, %s, %s)',
+                (log_date, name, data['up'], data['down'])
+            )
+
+
+def process_apache_log(projects, fn, log_date):
+    """Extract transfer data for a given day for all projects from an Apache log file.
+
+    We iteratively populate projects from all the log files in
+    consideration (4 by default) to avoid reading the entire file
+    repeatedly for multiple projects and to avoid having to merge
+    dictionaries.
+    """
+
+    with fn.open('r') as f:
+        for n, line in enumerate(f, 1):
+            stats = line.split()
+
+            # apache log date format looks like [11/Jul/2017:00:05:16 -0700]
+            # line.split()[3][1:12] = 11/Jul/2017
+            # we need to dump dates that don't match because
+            # logrotate rotates the logs at 6am
+            try:
+                line_date = datetime.strptime(stats[3][1:12], '%d/%b/%Y')
+            except ValueError:
+                print('Error on line {}:\\  {}\\'.format(n, line))
+                traceback.print_exc()
+
+            if line_date.date() != log_date:
+                continue
+
+            # extract dist name from request url
+            # '/debian/pool/main/h/hwdata/...' -> 'debian'
+            dist = stats[6]
+            dist = dist.split('/')[1] if '/' in dist else dist
+
+            # record if we returned http 2xx/3xx
+            if stats[8][0] in ('2', '3') and dist in projects:
+                projects[dist]['up'] += int(stats[-2])
+                projects[dist]['down'] += int(stats[-1])
+            else:
+                projects['other']['up'] += int(stats[-2])
+                projects['other']['down'] += int(stats[-1])
+
+    return projects
+
+
+def process_rsync_log(filename):
+    """Parse Tx/Rx information out of an rsyncd log file.
+
+    This function is project-agnostic.
+    """
+
+    tx = 0
+    rx = 0
+    with filename.open('r') as f:
+        for line in f.readlines():
+            # log format is <datetime> <pid> sent <tx> bytes received <rx> bytes total size
+            if line.find('sent') > 0:
```

Comment:
hopefully we never mirror a project called sent

Reply:
Like @ethanhs said above, I think a regex would also be cleaner here.